[libc++][hardening] Allow setting the assertion semantic via CMake.

var-const opened the pull request:
Add a new CMake variable, `LIBCXX_ASSERTION_SEMANTIC`, that largely
mirrors `LIBCXX_HARDENING_MODE`, except that it also supports a
special value `hardening_dependent` that indicates the semantic will be
selected based on the hardening mode in effect:
- `fast` and `extensive` map to `quick_enforce`;
- `debug` maps to `enforce`.

var-const committed:
[libc++][hardening] Allow setting the assertion semantic via CMake.

Add a new CMake variable, `LIBCXX_ASSERTION_SEMANTIC`, that largely
mirrors `LIBCXX_HARDENING_MODE`, except that not setting it is valid and
results in the default mapping from the selected hardening mode:
- `fast` and `extensive` map to `quick_enforce`;
- `debug` maps to `enforce`.

The default assertion semantic set via CMake takes precedence over the
default mapping but can still be overridden by setting the
`_LIBCPP_ASSERTION_SEMANTIC` macro directly.

var-const committed:
Fix the check for empty string in `CMakeLists.txt`.

philnik777 reviewed (commented):
Could you fix the formatting? Right now I have no idea what the actual change is.

var-const committed:
Fix formatting.

var-const commented:
> Could you fix the formatting? Right now I have no idea what the actual change is.

Done.

ldionne commented on libcxx/include/__configuration/hardening.h:
```diff
@@ -156,26 +156,47 @@ _LIBCPP_HARDENING_MODE_DEBUG
-
+// If the user or the vendor attempt to configure the assertion semantic, check that it is allowed in the current
+// environment.
+#if defined(_LIBCPP_ASSERTION_SEMANTIC) || defined(_LIBCPP_ASSERTION_SEMANTIC_DEFAULT)
```
`_LIBCPP_ASSERTION_SEMANTIC_DEFAULT` will always be defined. In fact, it is currently always defined. So I don't understand why we aren't hitting

```
#  if defined(_LIBCPP_CXX03_LANG)
#    error "Assertion semantics are not available in the C++03 mode."
#  endif
```

in the CI.

var-const replied:
I think it wouldn't be defined if not set by the "vendor" in the original version of the patch, but it's a moot point now.

ldionne commented on libcxx/include/__configuration/hardening.h:
```diff
@@ -156,26 +156,47 @@ _LIBCPP_HARDENING_MODE_DEBUG
+// (i.e., trap) and the `debug` mode maps to `enforce` (i.e., log and abort).
+#ifndef _LIBCPP_ASSERTION_SEMANTIC // User-provided semantic takes top priority -- don't override if set.
+
+#  ifdef _LIBCPP_ASSERTION_SEMANTIC_DEFAULT // Vendor-provided semantic takes second priority.
```
I suggest we always define `_LIBCPP_ASSERTION_SEMANTIC_DEFAULT`, like we do for hardening:

```
#  ifndef _LIBCPP_HARDENING_MODE_DEFAULT
#    error _LIBCPP_HARDENING_MODE_DEFAULT is not defined. This definition should be set at configuration time in the `__config_site` header, please make sure your installation of libc++ is not broken.
#  endif
```

ldionne commented on libcxx/include/__configuration/hardening.h:
```diff
@@ -156,26 +156,47 @@ _LIBCPP_HARDENING_MODE_DEBUG
+_LIBCPP_ASSERTION_SEMANTIC_QUICK_ENFORCE, \
+_LIBCPP_ASSERTION_SEMANTIC_ENFORCE
+#endif
+
```
We should add some tests for this.

var-const committed:
- Add new tests to check that the choice of the assertion semantic is
  orthogonal to the selected hardening mode;
- Tweak some checks in existing tests to rely on the assertion semantic
  rather than using the hardening mode as a proxy.

var-const committed:
Make `hardening_dependent` a special value for choosing the assertion
semantic (rather than relying on it being not set).

var-const commented on libcxx/include/__configuration/hardening.h:
```diff
@@ -150,32 +162,53 @@ _LIBCPP_HARDENING_MODE_DEBUG
-#  define _LIBCPP_ASSERTION_SEMANTIC_OBSERVE       (1 << 2)
-#  define _LIBCPP_ASSERTION_SEMANTIC_QUICK_ENFORCE (1 << 3)
-#  define _LIBCPP_ASSERTION_SEMANTIC_ENFORCE       (1 << 4)
+#  define _LIBCPP_ASSERTION_SEMANTIC_HARDENING_DEPENDENT (1 << 1)
```
Are we happy with these values? Should `hardening-dependent` be last, perhaps?

var-const commented on libcxx/include/__configuration/hardening.h:
```diff
@@ -135,13 +135,25 @@ _LIBCPP_HARDENING_MODE_EXTENSIVE, \
 // - `quick-enforce` terminates the program as fast as possible (via trapping);
 // - `enforce` logs an error and then terminates the program.
 //
+// Additionally, a special `hardening-dependent` value selects the assertion semantic based on the hardening mode in
```
Another name I had in mind was `derive-from-hardening`, let me know if you have a preference or any other suggestions.

var-const committed:
CI fixes (formatting, disable semantic-related tests in the C++03 mode)

var-const committed:
Fix CI (formatting, make semantics tests require experimental)

ldionne commented on libcxx/test/libcxx/assertions/semantics/override_with_enforce_semantic.pass.cpp:
```diff
@@ -0,0 +1,29 @@
+//
+//===----------------------------------------------------------------------===//
+
+// This test ensures that we can override the assertion semantic used by any checked hardening mode with `enforce` (this
```
I think we should. We don't necessarily have a need for it right now, but it seems confusing to have two similar settings (hardening mode/assertion semantics) not offering the same flexibility.

var-const commented on libcxx/include/__configuration/hardening.h:
```diff
@@ -135,13 +135,25 @@ _LIBCPP_HARDENING_MODE_EXTENSIVE, \
 //
+// Additionally, a special `hardening-dependent` value selects the assertion semantic based on the hardening mode in
+// effect: the production-capable modes (`fast` and `extensive`) map to `quick_enforce` and the `debug` mode maps to
+// `enforce`.
```
Thanks!

var-const committed:
- Reflect the assertion semantic in the library ABI tag (with a test);
- Address other feedback.

var-const commented on libcxx/include/__config:
```diff
@@ -291,14 +291,26 @@ typedef __char32_t char32_t;
 #  define _LIBCPP_ODR_SIGNATURE                                                                                        \
-    _LIBCPP_CONCAT(_LIBCPP_CONCAT(_LIBCPP_HARDENING_SIG, _LIBCPP_EXCEPTIONS_SIG), _LIBCPP_VERSION)
+    _LIBCPP_CONCAT(                                                                                                    \
+        _LIBCPP_CONCAT(_LIBCPP_CONCAT(_LIBCPP_HARDENING_SIG, _LIBCPP_ASSERTION_SEMANTIC_SIG), _LIBCPP_EXCEPTIONS_SIG), \
```
Do we want to eventually add a 3-argument version of `_LIBCPP_CONCAT` (maybe even 4-argument)?

var-const commented on libcxx/include/__config:
```diff
@@ -291,14 +291,26 @@ typedef __char32_t char32_t;
+#    define _LIBCPP_ASSERTION_SEMANTIC_SIG q
+#  elif _LIBCPP_ASSERTION_SEMANTIC == _LIBCPP_ASSERTION_SEMANTIC_ENFORCE
+#    define _LIBCPP_ASSERTION_SEMANTIC_SIG e
+#  else
```
Very minor point, but I'm a little on the fence about not checking for the `ignore` semantic explicitly. I ended up leaning towards this way to avoid duplicating the check for a valid semantic which IMO doesn't seem like it belongs here.

ldionne reviewed (approved):
LGTM with comments applied and CI passing!

var-const committed:
Address feedback: update documentation, marka test unsupported in C++03.

var-const committed:
Fix the check for C++03 in a test

var-const committed:
Mark a test as unsupported, take 3
